### PR TITLE
Fix/server signal refactor

### PR DESCRIPTION
**Pull request by jenr24G**
Refactors the signal processing code from the scope of `construct_responses()` down to be more readable and testable

### CODE REVIEW

**Review by mxnkarou — CHANGES_REQUESTED** [review]
Much more readable - good stuff!
One issue came up - would be great if you could check that out:
It seems like the sorting is not working. Ideally we'd want the signals sorted by timestamp.

**Comment by jenr24G** [comment]
An antipattern I noticed throughout this refactor, using Option's for delayed initialization. initializing a value to None so that you can have a draft object in state to be finished later makes all the code that has to deal with that Option spaghetti-like and just a plain mess. Use Option's for potential failure, not delayed initialization plz thx bye

**Review by mxnkarou — CHANGES_REQUESTED** [review]
Sorting looks good!
One thing we'd need to revert is the folder/file structure to reflect the actual endpoints.
Common practice when building REST API's is to keep routes in sync with files.
i.E. in this case the route that is currently handled in `router/api/mod.rs` should be `router/api/keyword.rs`.


And concerning the first comment - let's keep counting settled/unsettled votes.

File: server/src/processor.rs [hunk 1]
```diff
@@ -213,17 +214,22 @@ impl <'a> SignalProcessor<'a> {
...
+        }
+
         ResponseEntity {
             settled,
             unsettled: self.unsettled_signals,
             invalid: self.invalid_signals,
-            stats
+            stats: stats_option
         }
     }
```
Comment (mxnkarou) [inline server/src/processor.rs:234]: You can match directly on the `ResponseEntity` struct without having to shadow `None` with `stats_option`.

`fn generate_response(self) -> ResponseEntity {
        let stats = self.stats();
        let settled = self
            .current_settled
            .into_iter()
            .map(|(_, v)| v)
            .collect::<Vec<Signal>>();

        ResponseEntity {
            settled,
            unsettled: self.unsettled_signals,
            invalid: self.invalid_signals,
            stats: match stats.yes != 0. && stats.no != 0. {
                true => Some(stats),
                false => None,
            },
        }
    }`

Comment (mxnkarou) [reply server/src/processor.rs:234]: And another change would be required on `models/mod.rs - line 73`:
`self.stats = stats` => ` self.stats = Some(stats)`

File: server/src/models/mod.rs [hunk 2]
```diff
@@ -70,7 +70,7 @@ impl ResponseEntity {
     }
 
     pub fn with_stats(mut self, stats: SignalStats) -> Self {
-        self.stats = Some(stats);
+        self.stats = stats;
```
Comment (mxnkarou) [inline server/src/models/mod.rs:73]: This needs to be `self.stats = Some(stats);`

File: server/src/processor.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            stats_option = Some(stats);
+        }
+
+        ResponseEntity {
+            settled,
+            unsettled: self.unsettled_signals,
+            invalid: self.invalid_signals,
+            stats: stats_option
+        }
+    }
```
Comment (mxnkarou) [inline server/src/processor.rs:234]: You can match directly on the struct without shadowing `None` in `stats_option`:
```
fn generate_response(self) -> ResponseEntity {
        let stats = self.stats();
        let settled = self
            .current_settled
            .into_iter()
            .map(|(_, v)| v)
            .collect::<Vec<Signal>>();

        ResponseEntity {
            settled,
            unsettled: self.unsettled_signals,
            invalid: self.invalid_signals,
            stats: match stats.yes != 0. && stats.no != 0. {
                true => Some(stats),
                false => None,
            },
        }
    }

**Review by mxnkarou — CHANGES_REQUESTED** [review]

**Review by mxnkarou — APPROVED** [review]

File: client/src/types/components.d.ts [hunk 4]
```diff
@@ -20,7 +20,7 @@ export interface TableProps {
   data: SignalEntity[];
   query: string | undefined;
   isLoading: boolean;
-  stats: StatsEntity;
+  stats: StatsEntity | null;
```
Comment (jenr24G) [inline client/src/types/components.d.ts:23]: 😏​

Comment (jenr24G) [inline client/src/types/components.d.ts:23]: 😒​

**Pull request merged**